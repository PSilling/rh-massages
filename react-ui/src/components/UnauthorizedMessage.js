// react imports
import React, { Component } from 'react';

// util imports
import _t from '../utils/Translations.js';

/**
 * Message shown to unauthorized users.
 */
class UnauthorizedMessage extends Component {

  render() {
    return(
      <div>
        <h1>
          { this.props.title }
        </h1>
        <hr />
          <h2>
            { _t.translate('Unauthorized') }
          </h2>
      </div>
    )
  }
}

export default UnauthorizedMessage
