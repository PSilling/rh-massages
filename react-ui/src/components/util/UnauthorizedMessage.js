// react imports
import React from "react";
import PropTypes from "prop-types";

// util imports
import _t from "../../util/Translations";

/**
 * Simple title message shown to unauthorized users.
 */
const UnauthorizedMessage = function UnauthorizedMessage(props) {
  return (
    <div className="mt-3">
      <h2>{props.title}</h2>
      <hr />
      <h3>{_t.translate("Unauthorized")}</h3>
    </div>
  );
};

UnauthorizedMessage.propTypes = {
  /** title shown above unauthorized notification */
  title: PropTypes.string
};

UnauthorizedMessage.defaultProps = {
  title: "Unauthorized"
};

export default UnauthorizedMessage;
