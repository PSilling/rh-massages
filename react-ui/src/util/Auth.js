// Auth handler using Keycloak
import Keycloak from 'keycloak-js';

// util imports
import Util from './Util';

/**
 * Keycloak authorization holder. Functions called on Auth work with the Keycloak
 * conncetion variable.
 */
var Auth = function() {};

Auth.keycloak = Keycloak();

/**
 * Register new and update old Clients after Keycloak authorization success.
 */
Auth.keycloak.onAuthSuccess = () => {
  Util.get(Util.CLIENTS_URL + 'my/subscribed', (json) => {
    Auth.subscribed = json;
    Util.put(Util.CLIENTS_URL, Auth.getClient(), () => {
    }, false);
  });
}

Auth.subscribed = true;

Auth.isAuthenticated = () => {
  return Auth.keycloak.authenticated;
}

Auth.isAdmin = () => {
  return Auth.keycloak.hasRealmRole("admin");
}

Auth.getToken = () => {
  return Auth.keycloak.token;
}

Auth.getSub = () => {
  return Auth.keycloak.subject;
}

/**
 * Generates a new Client based on server Client representation.
 */
Auth.getClient = () => {
  return {
    sub: Auth.keycloak.subject, email: Auth.keycloak.idTokenParsed.email,
    name: Auth.keycloak.idTokenParsed.given_name, surname: Auth.keycloak.idTokenParsed.family_name,
    subscribed: Auth.subscribed
  };
}

export default Auth
