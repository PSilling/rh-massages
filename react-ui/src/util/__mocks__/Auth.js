const isAdmin = jest.fn().mockReturnValue(true);
const isAuthenticated = jest.fn().mockReturnValue(true);
const subscribed = true;
const getToken = jest.fn().mockReturnValue("secret");
const getSub = jest.fn().mockReturnValue("test");
const getClient = jest.fn().mockReturnValue({
  sub: "test", email: "test@email.com",
  name: "Test", surname: "User",
  subscribed: true
});

module.exports = {
  isAdmin,
  isAuthenticated,
  subscribed,
  getToken,
  getSub,
  getClient
}
