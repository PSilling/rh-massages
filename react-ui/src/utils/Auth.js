// Auth "service"
const Auth = {

  isAuthenticated() {
    return true;
  },

  isAdmin() {
    return true;
  }
}

export default Auth
