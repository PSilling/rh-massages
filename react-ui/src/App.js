// react imports
import React, { Component } from "react";
import { BrowserRouter as Router, Route, Link, Switch, withRouter } from "react-router-dom";
import PropTypes from "prop-types";

// module imports
import { Navbar, NavbarBrand, Nav, NavItem, NavLink } from "reactstrap";
import moment from "moment";
import { NotificationContainer } from "react-notifications";
import "bootstrap/dist/css/bootstrap.min.css";
import "react-notifications/lib/notifications.css";
import "react-datetime/css/react-datetime.css";
import "react-big-calendar/lib/css/react-big-calendar.css";

// view imports
import Facilities from "./views/Facilities";
import Massages from "./views/Massages";
import MassagesArchive from "./views/MassagesArchive";
import MyMassages from "./views/MyMassages";
import Settings from "./views/Settings";
import Users from "./views/Users";

// component imports
import ErrorBoundary from "./components/util/ErrorBoundary";
import ProfileLink from "./components/links/ProfileLink";
import LangLink from "./components/links/LangLink";
import LogoutLink from "./components/links/LogoutLink";
import UnauthorizedMessage from "./components/util/UnauthorizedMessage";

// util imports
import _t from "./util/Translations";
import Auth from "./util/Auth";
import Fetch from "./util/Fetch";

// moment.js localization
moment.updateLocale("en", {
  months: _t
    .translate("January_February_March_April_May_June_July_August_September_October_November_December")
    .split("_"),
  monthsShort: _t.translate("Jan_Feb_Mar_Apr_May_Jun_Jul_Aug_Sep_Oct_Nov_Dec").split("_"),
  weekdays: _t.translate("Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday").split("_"),
  weekdaysShort: _t.translate("Sun_Mon_Tue_Wed_Thu_Fri_Sat").split("_"),
  weekdaysMin: _t.translate("Su_Mo_Tu_We_Th_Fr_Sa").split("_"),
  longDateFormat: {
    LT: _t.translate("h:mm A"),
    L: _t.translate("DD/MM/YYYY")
  },
  week: {
    dow: _t.translate("7")
  }
});

/**
 * 404 page displayed when no route is matched.
 */
const NoMatch = ({ location }) => (
  <div className="text-center mt-5">
    <h4>
      <code>
        Error 404
        {_t.translate(" (page doesn't exist): ")}
        {location.pathname}
      </code>
    </h4>
    <h4 className="mt-4">
      <Link style={{ color: "#595959" }} to="/">
        {_t.translate("Back to main page")}
      </Link>
    </h4>
  </div>
);

/**
 * Application header.
 */
export const NavWithLinks = withRouter(({ location }) => (
  <Navbar color="light" light expand="md" className="no-print">
    <NavbarBrand>{_t.translate("Massages")}</NavbarBrand>
    <Nav navbar>
      <NavItem active={location.pathname === "/my-massages"}>
        <NavLink tag={Link} to="/my-massages" className="mr-1">
          {_t.translate("My Massages")}
        </NavLink>
      </NavItem>
      <NavItem active={location.pathname === "/"}>
        <NavLink tag={Link} to="/" className="mr-1">
          {_t.translate("Massages")}
        </NavLink>
      </NavItem>
      {Auth.isAdmin() && (
        <NavItem active={location.pathname === "/facilities"}>
          <NavLink tag={Link} to="/facilities" className="mr-1">
            {_t.translate("Facilities")}
          </NavLink>
        </NavItem>
      )}
      {Auth.isAdmin() && (
        <NavItem active={location.pathname === "/users"}>
          <NavLink tag={Link} to="/users" className="mr-1">
            {_t.translate("Users")}
          </NavLink>
        </NavItem>
      )}
      {Auth.isAdmin() && (
        <NavItem active={location.pathname === "/massages-archive"}>
          <NavLink tag={Link} to="/massages-archive" className="mr-1">
            {_t.translate("Massages Archive")}
          </NavLink>
        </NavItem>
      )}
      <NavItem active={location.pathname === "/settings"}>
        <NavLink tag={Link} to="/settings" className="mr-1">
          {_t.translate("Settings")}
        </NavLink>
      </NavItem>
    </Nav>
    <Nav navbar style={{ marginLeft: "auto", marginRight: "0" }}>
      <NavItem className="mr-1">
        <ProfileLink />
      </NavItem>
      <NavItem className="mr-1">
        <LogoutLink />
      </NavItem>
      <NavItem>
        <LangLink />
      </NavItem>
    </Nav>
  </Navbar>
));

/**
 * Main application component. Contains the Router, NotificationContainer and Navbar.
 */
class App extends Component {
  componentDidMount() {
    Fetch.tryWebSocketSend("ADD_Client");
  }

  componentWillUnmount() {
    Fetch.WEBSOCKET_CALLBACKS.client = null;
    Fetch.tryWebSocketSend("REMOVE_Client");

    if (Fetch.WEBSOCKET.readyState === WebSocket.OPEN) {
      Fetch.WEBSOCKET.close();
    }
  }

  render() {
    if (!Auth.isAuthenticated()) {
      return <UnauthorizedMessage title={_t.translate("Massages")} />;
    }

    return (
      <Router>
        <div>
          <NotificationContainer />

          <NavWithLinks />

          <div className="container">
            <ErrorBoundary>
              <Switch>
                <Route exact path="/" component={Massages} />
                <Route exact path="/my-massages" component={MyMassages} />
                <Route exact path="/facilities" component={Facilities} />
                <Route exact path="/users" component={Users} />
                <Route exact path="/massages-archive" component={MassagesArchive} />
                <Route exact path="/settings" component={Settings} />
                <Route component={NoMatch} />
              </Switch>
            </ErrorBoundary>
          </div>
        </div>
      </Router>
    );
  }
}

NoMatch.propTypes = {
  /** current router location */
  location: PropTypes.shape({
    pathname: PropTypes.string
  }).isRequired
};

export default App;
