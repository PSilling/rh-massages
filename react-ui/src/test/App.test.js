// react imports
import React from 'react';
import { Router, Link } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';

// test imports
import App, { NavWithLinks } from '../App';
import Auth from '../util/Auth';
import UnauthorizedMessage from '../components/util/UnauthorizedMessage';
import _t from '../util/Translations';

// test mocks
jest.mock('../util/Auth');
jest.mock('../util/Util');

afterEach(() => {
  jest.resetAllMocks();
});

test('renders admin content correctly', () => {
  const testRenderer = TestRenderer.create(<App />),
        testInstance = testRenderer.root;

  let router = testInstance.findByType(Router),
      links = testInstance.findAllByType(Link),
      rightBar = testInstance.findByProps({ className: "nav navbar-nav navbar-right" }),
      massagesLinks = testInstance.findAllByProps({ to: "/" }),
      myMassageLink = testInstance.findByProps({ to: "/my-massages" }),
      facilitiesLink = testInstance.findByProps({ to: "/facilities" }),
      massagesArchiveLink = testInstance.findByProps({ to: "/massages-archive" }),
      settingsLink = testInstance.findByProps({ to: "/settings" }),
      treeJSON = testRenderer.toJSON();

  expect(Auth.isAdmin).toHaveBeenCalledTimes(3);
  expect(Auth.isAuthenticated).toHaveBeenCalledTimes(3);
  expect(links).toHaveLength(6);
  expect(massagesLinks).toHaveLength(2);
  expect(massagesLinks[0].props.children).toBe(_t.translate("Massages"));
  expect(massagesLinks[1].props.children).toBe(_t.translate("Massages"));
  expect(myMassageLink.props.children).toBe(_t.translate("My Massages"));
  expect(facilitiesLink.props.children).toBe(_t.translate("Facilities"));
  expect(massagesArchiveLink.props.children).toBe(_t.translate("Massages Archive"));
  expect(settingsLink.props.children).toBe(_t.translate("Settings"));
  expect(rightBar.props.children).toHaveLength(3);
  expect(treeJSON).toMatchSnapshot();
});

test('renders unauthorized content correctly', () => {
  Auth.isAuthenticated = jest.fn().mockImplementation(() => false);

  const testRenderer = TestRenderer.create(<App />),
        testInstance = testRenderer.root;

  let message = testInstance.findByType(UnauthorizedMessage),
      treeJSON = testRenderer.toJSON();

  expect(Auth.isAdmin).not.toHaveBeenCalled();
  expect(Auth.isAuthenticated).toHaveBeenCalledTimes(1);
  expect(message.props.title).toBe(_t.translate('Massages'));
  expect(treeJSON).toMatchSnapshot();
});

test('hides non-admin Links correctly', () => {
  Auth.isAuthenticated = jest.fn().mockImplementation(() => true);
  Auth.isAdmin = jest.fn().mockImplementation(() => false);

  const testRenderer = TestRenderer.create(<App />),
        testInstance = testRenderer.root;

  let links = testInstance.findAllByType(Link);

  expect(Auth.isAdmin).toHaveBeenCalledTimes(3);
  expect(Auth.isAuthenticated).toHaveBeenCalledTimes(2);
  expect(links).toHaveLength(4);
});
