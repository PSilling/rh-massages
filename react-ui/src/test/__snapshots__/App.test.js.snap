// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders admin content correctly 1`] = `
<div>
  <div
    className="notification-container notification-container-empty"
  >
    <span />
  </div>
  <nav
    className="navbar navbar-default no-print"
  >
    <div
      className="container-fluid"
    >
      <div
        className="navbar-header"
      >
        <a
          className="navbar-brand"
          href="/"
          onClick={[Function]}
        >
          Masáže
        </a>
      </div>
      <ul
        className="nav navbar-nav"
      >
        <li
          className=""
        >
          <a
            href="/my-massages"
            onClick={[Function]}
          >
            Moje masáže
          </a>
        </li>
        <li
          className="active"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Masáže
          </a>
        </li>
        <li
          className=""
        >
          <a
            href="/facilities"
            onClick={[Function]}
          >
            Budovy
          </a>
        </li>
        <li
          className=""
        >
          <a
            href="/massages-archive"
            onClick={[Function]}
          >
            Archiv masáží
          </a>
        </li>
        <li
          className=""
        >
          <a
            href="/settings"
            onClick={[Function]}
          >
            Nastavení
          </a>
        </li>
      </ul>
      <ul
        className="nav navbar-nav navbar-right"
      >
        <li>
          <button
            className="btn btn-link navbar-btn"
            onClick={[Function]}
            title="Zobrazit Keycloak profil"
            type="button"
          >
            <span
              className="glyphicon glyphicon-user"
            />
             
            Profil
          </button>
        </li>
        <li>
          <button
            className="btn btn-link navbar-btn"
            onClick={[Function]}
            type="button"
          >
            <span
              className="glyphicon glyphicon-log-out"
            />
             
            Odhlásit
          </button>
        </li>
        <li>
          <button
            className="btn btn-link navbar-btn"
            onClick={[Function]}
            title="Změnit jazyk na angličtinu"
            type="button"
          >
            <span
              className="glyphicon glyphicon-globe"
            />
             
            EN
          </button>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="container"
  >
    <div>
      <div
        className="no-print"
      >
        <div>
          <h1>
            Masáže
          </h1>
          <h3>
            Žádné
          </h3>
        </div>
      </div>
      
    </div>
  </div>
</div>
`;

exports[`renders unauthorized content correctly 1`] = `
<div>
  <h1>
    Masáže
  </h1>
  <hr />
  <h2>
    Neautorizováno
  </h2>
</div>
`;
